core/cache/lxdprofilewatcher.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package cache
+
+import (
+	"github.com/juju/collections/set"
+	"github.com/juju/pubsub"
+
+	"github.com/juju/juju/core/lxdprofile"
+)
+
+type MachineAppLXDProfileWatcher struct {
+	*notifyWatcherBase
+
+	applications map[string]appInfo // unit names for each application
+	machineId    string
+
+	getApp   appFunc
+	getCharm charmFunc
+	getUnit  unitFunc
+}
+
+type appFunc func(string) (*Application, error)
+type charmFunc func(string) (*Charm, error)
+type unitFunc func(string) (*Unit, error)
```

Comment:
I think the following three type functions could be replaced with one interface, that way it performs the same role as the three functions, but we can do more (mock the new type)...


Something vaguely like:
```go
// MachineAppModeler is a point of use model for MachineAppLXDProfileWatcher to
// get Applications, Charms and Units.
type MachineAppModeler interface {
    Application(string) (*Application, error)
    Charm(string) (*Charm, error)
    Unit(string) (*Unit, error)
}
```

Reply:
Good call, i kept adding over time, should have looked back!